spices/SPICE-0009-external-readers.adoc

```diff
@@ -0,0 +1,120 @@
+= External Readers
+
+* Proposal: link:./SPICE-0009-external-readers.adoc[SPICE-0009]
+* Status: TBD
+* Implemented in: Pkl 0.27 ()
+* Category: Language
+
+== Introduction
+
+This SPICE proposes a way for custom module and resource readers to be used when evaluating modules with `pkl eval`.
+
+== Motivation
+
+When used via language binding libraries, custom module and resource readers schemes may be implemented by client code.
+Custom readers help bridge Pkl into systems where it does't fit natively, enabling a wide variety of use cases:
+
+* Mediating access to resources that require authentication (eg. a secret store)
+* Querying data accessible via non-HTTP protocols (eg. LDAP)
+* Providing filesystem abstractions for remote or in-memory modules
+* And more!
+
+One of the primary drawbacks to custom readers is that using one instantly turns a Pkl module into a _incompatible language_: the module may only be evaoluated by the client tool or application providing the reader and can no longer be evaluated directly using `pkl eval`.
+This hinders development and debugging of such modules, as typical workflows involving `trace(<expr>)` or `pkl eval -x '<expr>'` behave differently or are no longer possible.
+
+== Proposed Solution
+
+The existing message passing API and client libraries will be expanded to support external readers.
+Instead of client libraries invoking `pkl server`, `pkl eval` invocations, when configured with an external reader, will run the reader executable as a subprocess.
+External readers can provide any number of resource and/or module schemes and will configured as an executable and list of arguments via:
+
+* CLI flag
+* PklProject property
+* Java API
+* Message passing API
+
+
+[source,pkl]
+----
+/// May be specified as an URI to a readable resource, eg. `file:`, `https:`, or `package:`
+/// May also be specified as just an executable name, in which case it will be resolved according to the PATH environment variable
+executable: String
```

Comment:
How would a URI work here? Like, how would `executable = "file:///path/to/foo"` work?

Reply:
My main hope with supporting URIs is that external reader binaries could be packaged directly in Pkl packages. A PklProject using such a reader might look something like this:
```pkl
import "pkl:platform"

dependencies {
  ["ldap"] {
    uri = "package://...@x.y.z"
  }
}

evaluatorSettings {
  allowedResources {
    "ldap:"
    "ldaps:"
  }
  externalReaders {
    new {
      executable = "@ldap/reader_\(platform.current.operatingSystem.name)_\(platform.current.processor.architecture)"
      // support for project dependency notation would be ideal!
    }
  }
}
```

Pkl would need to know how to translate `file:` URIs to absolute paths and for `https:`, `package:`, or `projectpackage:` URIs would download the file/package and execute the reader from some tmp/cache path.

Enabling this workflow would avoid complicating distribution of external reader binaries and would allow packages to provide nicer (not stringly-typed) APIs for their readers. For example, the `ldap:` example in the proposal requires detailed understanding of [RFC 4516](https://docs.ldap.com/specs/rfc4516.txt). It might be nicer if an LDAP Pkl package existed that provided an API like this:
```pkl
protocol: *"ldap"|"ldaps"
host: String
port: UInt16 = 389
baseDN: String
attributes: Listing<String>
scope: *"base"|"one"|"sub"
filter: String = "(&)" // matches anything

typealias Result = Mapping<String, Listing<String>>

fixed results: Listing<Result> = new json.Parser { useMapping = true }.parse(
  read("\(protocol)://\(host):\(port)/\(baseDN)?\(attributes.join(","))?\(scope)?\(filter)").text
)
```

Reply:
note: see comments [here](https://github.com/apple/pkl-evolution/pull/10#pullrequestreview-2227333322) and [here](https://github.com/apple/pkl-evolution/pull/10#issuecomment-2278850140) about why not distribute binaries using a package